added set_audio_output() and updated readme

Felix-Pi opened the pull request.

supersaiyanmode commented:
Hi there, Thanks so much for this PR!

I have a few questions:

1. Is there a "get_audio_output()" equivalent API?
2. Can we subscribe to it?

Additionally, as for the actual code changes, I was wondering turning those string into an enum? We don't want to the users to pass in the actual payload string. The best option is to introduce a simple class in `pywebstv.models`, and use it like `ApplicationControl` or `SourceControl`.

Felix-Pi commented:
Hi, thanks for your quick reply and your input!

I’ve added get_audio_output() which can be subscribed to. 
As far as I know there is no endpoint like 'listAudioSources‘, so I implemented a method list_audio_output_sources() to mimic that and return a list of instances from AudioOutputSource.
set_audio_output(..) now accepts an instance of AudioOutputSource.

supersaiyanmode commented on pywebostv/controls.py:
```diff
@@ -176,6 +181,20 @@ class MediaControl(WebOSControlBase):
+        "get_audio_output": {
+            "uri": "ssap://audio/getStatus",
+            "validation": standard_validation,
+            "subscription": True,
```
```suggestion
            "subscription": True,
            "return": AudioOutputSource
```

That should convert the incoming response from LG TV Server to an AudioOutputSource instance. This is untested, though :)

If this request returns a more complex object, you'd have to do something like ApplicationControl -> list_apps -> "return" (line number 248 below).

supersaiyanmode replied:
Also, can you please dump the entire response as you get from the TV? There might be some other interesting fields..

Felix-Pi replied:
its just {'soundOutput': 'external_speaker'}

supersaiyanmode replied:
Ahh, thanks so much. Just a few things to note in that case:

You might want to remove `"standard_validation"` then (the entire key+value). I am guessing it would fail because of this line: https://github.com/supersaiyanmode/PyWebOSTV/blob/master/pywebostv/controls.py#L117?

supersaiyanmode replied:
What is the response if you print out the object here: https://github.com/supersaiyanmode/PyWebOSTV/blob/master/pywebostv/connection.py#L194 ?

Felix-Pi replied:
**received_message() output:** 
{"type":"response","id":"5a1a03b3-03d1-4264-b83e-cf89be987db0","payload":{"returnValue":true,"soundOutput":"external_speaker"}}

**Output with standard_validation:** 
{'soundOutput': 'external_speaker'}

**Output without standard_validation:** 
{'returnValue': True, 'soundOutput': 'external_speaker'}

So I guess it works, right?

supersaiyanmode replied:
Yes, it works with standard_validation :). Let's not remove standard_validation then

supersaiyanmode commented on pywebostv/controls.py:
```diff
@@ -153,6 +153,11 @@ def request_func():
 
 class MediaControl(WebOSControlBase):
+    def list_audio_output_sources(self):
+        sources = ['tv_speaker', 'external_speaker', 'soundbar', 'bt_soundbar', 'tv_external_speaker']
```
I just quick google search with the URI you discovered. That led me here: https://github.com/bendavid/aiopylgtv/issues/3#issuecomment-575914925. Turns out there could potentially be more than the 5 items you have in your list. This is why I was curious if we could get this list from the TV itself rather than hard-coding anything. Can you point me to the place where you found this URI (also was it documented somewhere) ?

Felix-Pi replied:
The link you found is where I got this from. Searched quiet a while for a possibility to get the list from the TV itself but could not find anything. So I ended up using the sources I could test with my own TV ('tv_speaker', 'external_speaker', 'soundbar', 'bt_soundbar', 'tv_external_speaker').

supersaiyanmode replied:
I think using a hard-coded list like this might lead to issues. What if I use "bt_soundbar" and I don't have a sound bar at all.

Further, in such a case, what is the response of TV when I set_audio_output(..) with that value? If it gives us an error indication of some sort, we will have to confirm that standard_validation is able to take care of it. Or else, we might have to write our own custom validation.

Felix-Pi replied:
You're right, do you have any idea how we could get a list of possible sources from the TV? 

**Output using an existing value:** 
{"type":"response","id":"ec489afb-3d7a-493b-8233-4a65a7dffbe9","payload":{"returnValue":true,"soundOutput":"external_speaker"}}

**Output using non existing value:**
  File "/usr/local/lib/python3.8/site-packages/pywebostv-0.8.4-py3.8.egg/pywebostv/controls.py", line 119, in request_func
OSError: There is No matched extended item: soundOutput

{"type":"error","id":"5a5caa03-9a3c-43f9-aefa-677a4664c07b","error":"500 Application error","payload":{"returnValue":false,"errorText":"There is No matched extended item: soundOutput"}}

supersaiyanmode replied:
On the brighter side, good to know the validation format is the same and it works.  I don't know how to list audio sources unfortunately.

supersaiyanmode commented on pywebostv/controls.py:
```diff
@@ -176,6 +181,20 @@ class MediaControl(WebOSControlBase):
         "stop": {"uri": "ssap://media.controls/stop"},
         "rewind": {"uri": "ssap://media.controls/rewind"},
         "fast_forward": {"uri": "ssap://media.controls/fastForward"},
+        "get_audio_output": {
```
Can you add this to the README.md as well?

Felix-Pi replied:
done :)

supersaiyanmode commented on pywebostv/model.py:
```diff
@@ -20,3 +20,14 @@ def __getitem__(self, val):
+
+    def __getitem__(self):
+        return self.data
+
```
If we don't have any interesting fields in the response (or in other words, we only care about the main string like "tv_output" etc), we can get rid of this entire `__getitem__`. This class doesn't have any dict-like access requirements.

Felix-Pi replied:
Okay, thanks again for your kind input! I've just recently started getting into python and looking forward to anything I can learn!

supersaiyanmode replied:
Ohh, you are super awesome for being able to navigate through the sources on your own esp when you're a beginner :)

Felix-Pi replied:
Thanks, really appreciate it!!

alliefitter commented:
@supersaiyanmode Whoa! There's a bunch of [stuff in here](https://github.com/merdok/homebridge-webos-tv/blob/master/lib/LgTvController.js) that isn't implement in this project. I'll try to implement and test all of this stuff this week and put in a PR. Do you still not have access to a WebOS TV to test stuff yourself? Man, I should have thought about looking at Home Bridge stuff.

supersaiyanmode commented:
> @supersaiyanmode Whoa! There's a bunch of [stuff in here](https://github.com/merdok/homebridge-webos-tv/blob/master/lib/LgTvController.js) that isn't implement in this project. I'll try to implement and test all of this stuff this week and put in a PR. Do you still not have access to a WebOS TV to test stuff yourself? Man, I should have thought about looking at Home Bridge stuff.

Totally! It is disappointing that LG refuses to release any sort of documentation. We have to go hunt for other uses in the wild, come back and implement it.

I have access to the TV, but the wireless receiver is non-functional for some reason. If I must test for some reason, it's via a 10ft long ethernet cable :stuck_out_tongue_closed_eyes:.

supersaiyanmode commented on README.md:
```diff
@@ -132,6 +132,10 @@ media.pause()
 media.rewind()
 media.fast_forward()
+
+media.get_audio_output()   #returns the currently used audio output source as AudioOutputSource instance
```
```suggestion
cur_media_output_source = media.get_audio_output()   # Returns the currently used audio output source as AudioOutputSource instance.
```

supersaiyanmode commented on README.md:
```diff
@@ -132,6 +132,10 @@ media.pause()
 media.fast_forward()
+
+media.get_audio_output()   #returns the currently used audio output source as AudioOutputSource instance
+audioOutputs = media.list_audio_output_sources()    # Returns a list of AudioOutputSource instances.
```
Suggestion to rename variable name for consistency.

```suggestion
audio_outputs = media.list_audio_output_sources()    # Returns a list of AudioOutputSource instances.
```

supersaiyanmode commented on README.md:
```diff
@@ -132,6 +132,10 @@ media.pause()
+
+media.get_audio_output()   #returns the currently used audio output source as AudioOutputSource instance
+audioOutputs = media.list_audio_output_sources()    # Returns a list of AudioOutputSource instances.
+media.set_audio_output(audioOutputs[0])             # .set_audio_output(..) accepts an AudioOutputSource instance.
```
```suggestion
media.set_audio_output(audio_outputs[0])             # .set_audio_output(..) accepts an AudioOutputSource instance.
```

supersaiyanmode reviewed (changes requested):
Looking good overall. Thanks so much for this PR! Just a tiny round of super minor requests :)

Felix-Pi commented:
> Looking good overall. Thanks so much for this PR! Just a tiny round of super minor requests :)

Thank you for your time and advise :)
